### PR TITLE
add-encryption-reconcilation-loop

### PR DESCRIPTION
**Pull request by calvix**
so this implement the basic function for the encryption provider it will do the following:

1) check if the encryption config already exists ( that means if the cluster has one) If not  create a new one
2a) if the cluster is completely new just generate a new encryption key and create the config that will be used
2b) if the cluster already exists but uses the old legacy encryption config from cluster-operator just migrate the key to the new scheme ( that will be used in the new release later)

Once the new encryption config exists we can check  for key rotation ( not implemented on this PR will be a separate new PR)

It's not provider-specific and only a slight change will be needed in the provider operators to use this system.
Should be easily used for both GiantSwarm legacy product and CAPI.

### CODE REVIEW

File: pkg/encryption/encryption.go [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+		s.logger.Error(err, "failed to get encryption provider config secret for cluster")
+		return err
+	} else {
+		// config already exists, check for key rotation
+		err = s.keyRotation(ctx)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
```
Comment (AverageMarcus) [inline pkg/encryption/encryption.go:105]:
```suggestion

	// config already exists, check for key rotation
	return s.keyRotation(ctx)
```

File: pkg/encryption/encryption.go [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+	err = s.ctrlClient.Create(ctx, encryptionProviderSecret)
+	if err != nil {
+		s.logger.Error(err, "failed to create encryption provider secret")
+		return err
+	}
+	return nil
+}
+
+func (s *Service) keyRotation(ctx context.Context) error {
+	// TODO implement the magic
```
Comment (AverageMarcus) [inline pkg/encryption/encryption.go:213]: Love it!

File: pkg/encryption/encryption.go [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+		}
+	}
+
+	secretData := map[string]string{}
+	// render template for the config
+	{
+		templateData := Data{
+			Keys: []Key{
+				{
+					Name:          "key1",
```
Comment (AverageMarcus) [inline pkg/encryption/encryption.go:170]: Does this name hold any special meaning?

File: pkg/label/label.go [hunk 4]
```diff
@@ -0,0 +1,14 @@
+package label
+
+const (
+	// Cluster label is a new style label for ClusterID
+	Cluster = "giantswarm.io/cluster"
+	// ManagedBy label denotes which operator manages corresponding resource.
+	ManagedBy = "giantswarm.io/managed-by"
```
Comment (AverageMarcus) [inline pkg/label/label.go:7]: Can we pull these two from https://github.com/giantswarm/k8smetadata/tree/main/pkg/label ?

**Review by AverageMarcus — CHANGES_REQUESTED** [review]
Just some small suggestions

File: pkg/encryption/encryption.go (hunk 3 above)
Comment (calvix) [reply pkg/encryption/encryption.go:170]: I can't say i understand the reason but they use this in the examples and value increment  by key index
https://kubernetes.io/docs/tasks/administer-cluster/encrypt-data/

File: pkg/label/label.go (hunk 4 above)
Comment (calvix) [reply pkg/label/label.go:7]: yeah good shout

File: pkg/encryption/encryption.go (hunk 1 above)
Comment (AverageMarcus) [reply pkg/encryption/encryption.go:105]: Ah, didn't realise there was more to go here eventually. Worth dropping in a `// TODO:` ?

**Review by AverageMarcus — APPROVED** [review]

**Pull request merged**